File: src/main/kotlin/com/procurement/access/service/CnOnPinServiceImpl.java
package com.procurement.access.service;

import com.procurement.access.model.dto.cn.CnLot;
import com.procurement.access.model.dto.cn.CnProcess;
import com.procurement.access.model.dto.cn.CnTender;
import com.procurement.access.model.dto.pin.PinLot;
import com.procurement.access.model.dto.pin.PinProcess;
import com.procurement.access.model.dto.pin.PinTender;
import com.procurement.access.model.entity.TenderProcessEntity;
import com.procurement.access.utils.DateUtil;
import com.procurement.access.utils.JsonUtil;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;
import org.springframework.stereotype.Service;

@Service
public class CnOnPinServiceImpl implements CnOnPinService {
    private final JsonUtil jsonUtil;
    private final DateUtil dateUtil;
    private final TenderProcessDao tenderProcessDao;


    public CnOnPinServiceImpl(final JsonUtil jsonUtil,
                              final DateUtil dateUtil,
                              final TenderProcessDao tenderProcessDao) {
        this.jsonUtil = jsonUtil;
        this.dateUtil = dateUtil;
        this.tenderProcessDao = tenderProcessDao;
    }

    @Override
    public ResponseDto createCnOnPin(final String cpId,
                                     final String previousStage,
                                     final String stage,
                                     final String owner,
                                     final String token,
                                     final LocalDateTime dateTime,
                                     final CnProcess cn) {

        final TenderProcessEntity entity = Optional.ofNullable(tenderProcessDao.getByCpIdAndStage(cpId, previousStage))
                .orElseThrow(() -> new ErrorException(ErrorType.DATA_NOT_FOUND));
        if (!entity.getOwner().equals(owner))
            throw new ErrorException(ErrorType.INVALID_OWNER);
        if (!entity.getToken().toString().equals(token))
            throw new ErrorException(ErrorType.INVALID_TOKEN);
        if (!entity.getCpId().equals(cn.getTender().getId()))
            throw new ErrorException(ErrorType.INVALID_CPID_FROM_DTO);
        final PinProcess pin = jsonUtil.toObject(PinProcess.class, entity.getJsonData());
        final PinTender pinTender = pin.getTender();

        if (!pinTender.getTenderPeriod().getStartDate().toLocalDate().equals(dateTime.toLocalDate()))
            throw new ErrorException(ErrorType.INVALID_START_DATE);

        /*planning*/
        cn.setPlanning(pin.getPlanning());
        /*tender*/
        final CnTender cnTender = convertPinToCnTender(pinTender);
        /*lots*/
        setLotsToCnFromPin(pinTender, cnTender);
         /*tender status, lot status*/
        setStatuses(cnTender);
        /*submissionLanguages*/
        if (cn.getTender().getSubmissionLanguages() != null)
            cnTender.setSubmissionLanguages(cn.getTender().getSubmissionLanguages());
        /*documents*/
        if (cn.getTender().getDocuments() != null)
            cnTender.setDocuments(cn.getTender().getDocuments());
        cn.setTender(cnTender);
        validateLots(cnTender);
        tenderProcessDao.save(getEntity(cn, cpId, stage, entity.getToken(), dateTime, owner));
        cn.setOcId(cpId);
        cn.setToken(entity.getToken().toString());
        return new ResponseDto<>(true, null, cn);
    }

    private void setLotsToCnFromPin(final PinTender pinTender, final CnTender cnTender) {
        if (pinTender.getLots() != null) {
            final List<CnLot> cnLots = pinTender.getLots().stream()
                    .map(this::convertPinToCnLot)
                    .collect(Collectors.toList());
            cnTender.setLots(cnLots);
        }
    }

    private void validateLots(final CnTender cnTender) {
        Set<String> lotsFromDocuments = null;
        if (cnTender.getDocuments() != null) {
            lotsFromDocuments = cnTender.getDocuments().stream()
                    .filter(document -> Objects.nonNull(document.getRelatedLots()))
                    .flatMap(d -> d.getRelatedLots().stream()).collect(Collectors.toSet());
        }

        if (cnTender.getLots() != null && lotsFromDocuments != null && !lotsFromDocuments.isEmpty()) {
            final Set<String> lotsFromCn = cnTender.getLots().stream().map(CnLot::getId).collect(Collectors.toSet());
            if (!lotsFromCn.containsAll(lotsFromDocuments))
                throw new ErrorException(ErrorType.INVALID_LOTS_RELATED_LOTS);
        }
    }

    private CnLot convertPinToCnLot(final PinLot pinLot) {
        return new CnLot(
                pinLot.getId(),
                pinLot.getTitle(),
                pinLot.getDescription(),
                pinLot.getStatus(),
                pinLot.getStatusDetails(),
                pinLot.getValue(),
                pinLot.getOptions(),
                pinLot.getRecurrentProcurement(),
                pinLot.getRenewals(),
                pinLot.getVariants(),
                pinLot.getContractPeriod(),
                pinLot.getPlaceOfPerformance()
        );
    }

    private CnTender convertPinToCnTender(final PinTender pinTender) {
        return new CnTender(
                pinTender.getId(),
                pinTender.getTitle(),
                pinTender.getDescription(),
                pinTender.getStatus(),
                pinTender.getStatusDetails(),
                pinTender.getClassification(),
                new HashSet(pinTender.getItems()),
                pinTender.getValue(),
                pinTender.getProcurementMethod(),
                pinTender.getProcurementMethodDetails(),
                pinTender.getProcurementMethodRationale(),
                pinTender.getMainProcurementCategory(),
                pinTender.getAdditionalProcurementCategories(),
                pinTender.getAwardCriteria(),
                pinTender.getSubmissionMethod(),
                pinTender.getSubmissionMethodDetails(),
                pinTender.getEligibilityCriteria(),
                pinTender.getContractPeriod(),
                pinTender.getProcuringEntity(),
                pinTender.getDocuments(),
                null,
                pinTender.getLotGroups(),
                pinTender.getAcceleratedProcedure(),
                pinTender.getDesignContest(),
                pinTender.getElectronicWorkflows(),
                pinTender.getJointProcurement(),
                pinTender.getLegalBasis(),
                pinTender.getProcedureOutsourcing(),
                pinTender.getProcurementMethodAdditionalInfo(),
                pinTender.getSubmissionLanguages(),
                pinTender.getSubmissionMethodRationale(),
                pinTender.getDynamicPurchasingSystem(),
                pinTender.getFramework(),
                pinTender.getRequiresElectronicCatalogue()
        );
    }

    private void setStatuses(CnTender cnTender) {
        cnTender.setStatus(TenderStatus.ACTIVE);
        cnTender.setStatusDetails(TenderStatusDetails.EMPTY);
        if (cnTender.getLots() != null)
            cnTender.getLots().forEach(lot -> {
                lot.setStatus(TenderStatus.ACTIVE);
                lot.setStatusDetails(TenderStatusDetails.EMPTY);
            });
    }

    private TenderProcessEntity getEntity(final CnProcess cn,
                                          final String cpId,
                                          final String stage,
                                          final UUID token,
                                          final LocalDateTime dateTime,
                                          final String owner) {
        final TenderProcessEntity entity = new TenderProcessEntity();
        entity.setCpId(cpId);
        entity.setToken(token);
        entity.setStage(stage);
        entity.setOwner(owner);
        entity.setCreatedDate(dateUtil.localToDate(dateTime));
        entity.setJsonData(jsonUtil.toJson(cn));
        return entity;
    }


}

File: src/main/kotlin/com/procurement/access/service/CnOnPinServiceImpl.kt
package com.procurement.access.service

import com.procurement.access.dao.TenderProcessDao
import com.procurement.access.model.dto.cn.CnLot
import com.procurement.access.model.dto.cn.CnProcess
import com.procurement.access.model.dto.cn.CnTender
import com.procurement.access.model.dto.ocds.TenderStatus
import com.procurement.access.model.dto.ocds.TenderStatusDetails
import com.procurement.access.model.dto.pin.PinLot
import com.procurement.access.model.dto.pin.PinProcess
import com.procurement.access.model.dto.pin.PinTender
import com.procurement.access.model.entity.TenderProcessEntity
import com.procurement.access.utils.DateUtil
import com.procurement.access.utils.JsonUtil
import com.procurement.access.utils.toDate
import com.procurement.notice.exception.ErrorException
import com.procurement.notice.exception.ErrorType
import com.procurement.notice.model.bpe.ResponseDto
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.util.*

interface CnOnPinService {

    fun createCnOnPin(
            cpId: String,
            previousStage: String,
            stage: String,
            owner: String,
            token: String,
            dateTime: LocalDateTime,
            data: CnProcess): ResponseDto<*>

}

@Service
class CnOnPinServiceImpl(private val jsonUtil: JsonUtil,
                         private val dateUtil: DateUtil,
                         private val tenderProcessDao: TenderProcessDao) : CnOnPinService {

    override fun createCnOnPin(cpId: String,
                               previousStage: String,
                               stage: String,
                               owner: String,
                               token: String,
                               dateTime: LocalDateTime,
                               cn: CnProcess): ResponseDto<*> {

        val entity = tenderProcessDao.getByCpIdAndStage(cpId, previousStage)
                ?: throw ErrorException(ErrorType.DATA_NOT_FOUND)
        if (entity.owner != owner) throw ErrorException(ErrorType.INVALID_OWNER)
        if (entity.token.toString() != token) throw ErrorException(ErrorType.INVALID_TOKEN)
        if (entity.cpId != cn.tender.id) throw ErrorException(ErrorType.INVALID_CPID_FROM_DTO)
        val pinProcess = jsonUtil.toObject(PinProcess::class.java, entity.jsonData)
        val pinTender = pinProcess.tender
        if (pinTender.tenderPeriod.startDate.toLocalDate() != dateTime.toLocalDate())
            throw ErrorException(ErrorType.INVALID_START_DATE)
        cn.planning = pinProcess.planning
        val cnTender = convertPinToCnTender(pinTender)
        setLotsToCnFromPin(pinTender, cnTender)
        setStatuses(cnTender)
        if (cn.tender.submissionLanguages != null) cnTender.submissionLanguages = cn.tender.submissionLanguages
        if (cn.tender.documents != null) cnTender.documents = cn.tender.documents
        cn.tender = cnTender
        validateLots(cnTender)
        tenderProcessDao.save(getEntity(cn, cpId, stage, entity.token, dateTime, owner))
        cn.ocId = cpId
        cn.token = entity.token.toString()
        return ResponseDto(true, null, cn)
    }

    private fun setLotsToCnFromPin(pinTender: PinTender, cnTender: CnTender) {
        if (pinTender.lots != null) {
            cnTender.lots = pinTender.lots.asSequence()
                    .map({ convertPinToCnLot(it) }).toList()
        }
    }

    private fun validateLots(cnTender: CnTender) {
        var lotsFromDocuments: HashSet<String>? = null
        if (cnTender.documents != null) {
            lotsFromDocuments = cnTender.documents!!.asSequence()
                    .filter({ it.relatedLots != null })
                    .flatMap({ it.relatedLots!!.asSequence() }).toHashSet()
        }
        if (cnTender.lots != null && lotsFromDocuments != null && !lotsFromDocuments.isEmpty()) {
            val lotsFromCn = cnTender.lots!!.asSequence()
                    .map({ it.id }).toHashSet()
            if (!lotsFromCn.containsAll(lotsFromDocuments)) throw ErrorException(ErrorType.INVALID_LOTS_RELATED_LOTS)
        }
    }

    private fun convertPinToCnLot(pinLot: PinLot): CnLot {
        return CnLot(
                id = pinLot.id,
                title = pinLot.title,
                description = pinLot.description,
                status = pinLot.status,
                statusDetails = pinLot.statusDetails,
                value = pinLot.value,
                options = pinLot.options,
                recurrentProcurement = pinLot.recurrentProcurement,
                renewals = pinLot.renewals,
                variants = pinLot.variants,
                contractPeriod = pinLot.contractPeriod,
                placeOfPerformance = pinLot.placeOfPerformance
        )
    }

    private fun convertPinToCnTender(pinTender: PinTender): CnTender {
        return CnTender(
                id = pinTender.id,
                title = pinTender.title,
                description = pinTender.description,
                status = pinTender.status,
                statusDetails = pinTender.statusDetails,
                classification = pinTender.classification,
                items = HashSet(pinTender.items),
                value = pinTender.value,
                procurementMethod = pinTender.procurementMethod,
                procurementMethodDetails = pinTender.procurementMethodDetails,
                procurementMethodRationale = pinTender.procurementMethodRationale,
                mainProcurementCategory = pinTender.mainProcurementCategory,
                additionalProcurementCategories = pinTender.additionalProcurementCategories,
                awardCriteria = pinTender.awardCriteria,
                submissionMethod = pinTender.submissionMethod,
                submissionMethodDetails = pinTender.submissionMethodDetails,
                eligibilityCriteria = pinTender.eligibilityCriteria,
                contractPeriod = pinTender.contractPeriod,
                procuringEntity = pinTender.procuringEntity,
                documents = pinTender.documents,
                lots = null,
                lotGroups = pinTender.lotGroups,
                acceleratedProcedure = pinTender.acceleratedProcedure,
                designContest = pinTender.designContest,
                electronicWorkflows = pinTender.electronicWorkflows,
                jointProcurement = pinTender.jointProcurement,
                legalBasis = pinTender.legalBasis,
                procedureOutsourcing = pinTender.procedureOutsourcing,
                procurementMethodAdditionalInfo = pinTender.procurementMethodAdditionalInfo,
                submissionLanguages = pinTender.submissionLanguages,
                submissionMethodRationale = pinTender.submissionMethodRationale,
                dynamicPurchasingSystem = pinTender.dynamicPurchasingSystem,
                framework = pinTender.framework,
                requiresElectronicCatalogue = pinTender.requiresElectronicCatalogue
        )
    }

    private fun setStatuses(cnTender: CnTender) {
        cnTender.status = TenderStatus.ACTIVE
        cnTender.statusDetails = TenderStatusDetails.EMPTY
        cnTender.lots?.forEach { lot ->
            lot.status = TenderStatus.ACTIVE
            lot.statusDetails = TenderStatusDetails.EMPTY
        }
    }

    private fun getEntity(cn: CnProcess,
                          cpId: String,
                          stage: String,
                          token: UUID,
                          dateTime: LocalDateTime,
                          owner: String): TenderProcessEntity {
        return TenderProcessEntity(
                cpId = cpId,
                token = token,
                stage = stage,
                owner = owner,
                createdDate = dateTime.toDate(),
                jsonData = jsonUtil.toJson(cn))
    }


}
